ouroboros-network/src/Ouroboros/Network/Protocol/Channel/Trace.hs

```diff
@@ -0,0 +1,29 @@
+module Ouroboros.Network.Protocol.Channel.Trace where
+
+import Cardano.BM.BaseTrace
+
+import Protocol.Channel
+
+-- TBD log before or after? Before seems better...
+traceSend
+  :: ( Applicative m, Functor n )
+  => BaseTrace m send
+  -> Duplex m n send recv
+  -> Duplex m n send recv
+traceSend bt duplex = duplex
+  { send = \item -> traceWith bt item *> fmap (traceSend bt) (send duplex item)
+  , recv = (fmap . fmap) (traceSend bt) (recv duplex)
+  }
```

Comment:
I think, alternatively you could use `channelSendEffect` and `channelRecvEffect`.

Reply:
Oh right I forgot about those.